File: Payload_Type/poseidon/mythic/agent_functions/sshauth.py
from mythic_payloadtype_container.MythicCommandBase import *
import json


class SSHAuthArguments(TaskArguments):
    def __init__(self, command_line, **kwargs):
        super().__init__(command_line, **kwargs)
        self.args = [
            CommandParameter(
                name="username",
                type=ParameterType.String,
                description="Authenticate to the designated hosts using this username.",
                parameter_group_info=[
                    ParameterGroupInfo(
                        ui_position=1,
                        group_name="scp-private-key"
                    ),
                    ParameterGroupInfo(
                        group_name="scp-plaintext-password"
                    ),
                    ParameterGroupInfo(
                        group_name="run-command-plaintext-password"
                    ),
                    ParameterGroupInfo(
                        group_name="run-command-private-key"
                    )
                ]
            ),
            CommandParameter(
                name="source",
                type=ParameterType.String,
                description="If doing SCP, this is the source file",
                default_value="",
                parameter_group_info=[
                    ParameterGroupInfo(
                        group_name="scp-private-key"
                    ),
                    ParameterGroupInfo(
                        group_name="scp-plaintext-password"
                    )
                ]
            ),
            CommandParameter(
                name="destination",
                type=ParameterType.String,
                description="If doing SCP, this is the destination file",
                default_value="",
                parameter_group_info=[
                    ParameterGroupInfo(
                        group_name="scp-private-key"
                    ),
                    ParameterGroupInfo(
                        group_name="scp-plaintext-password"
                    )
                ]
            ),
            CommandParameter(
                name="private_key",
                type=ParameterType.String,
                description="Authenticate to the designated hosts using this private key",
                parameter_group_info=[
                    ParameterGroupInfo(
                        group_name="scp-private-key"
                    ),
                    ParameterGroupInfo(
                        group_name="command-private-key"
                    )
                ]
            ),
            CommandParameter(
                name="port",
                type=ParameterType.Number,
                description="SSH Port if different than 22",
                default_value="22",
                parameter_group_info=[
                    ParameterGroupInfo(
                        required=False,
                        group_name="scp-private-key"
                    ),
                    ParameterGroupInfo(
                        required=False,
                        group_name="scp-plaintext-password"
                    ),
                    ParameterGroupInfo(
                        group_name="run-command-plaintext-password"
                    ),
                    ParameterGroupInfo(
                        group_name="run-command-private-key"
                    )
                ]
            ),
            CommandParameter(
                name="password",
                type=ParameterType.String,
                description="Authenticate to the designated hosts using this password",
                default_value="",
                parameter_group_info=[
                    ParameterGroupInfo(
                        group_name="scp-plaintext-password"
                    ),
                    ParameterGroupInfo(
                        group_name="command-plaintext-password"
                    )
                ]
            ),
            CommandParameter(
                name="hosts",
                type=ParameterType.Array,
                description="Hosts that you will auth to",
                parameter_group_info=[
                    ParameterGroupInfo(
                        group_name="scp-plaintext-password"
                    ),
                    ParameterGroupInfo(
                        group_name="scp-private-key"
                    ),
                    ParameterGroupInfo(
                        group_name="run-command-plaintext-password"
                    ),
                    ParameterGroupInfo(
                        group_name="run-command-private-key"
                    )
                ]
            ),
            CommandParameter(
                name="command",
                type=ParameterType.String,
                description="Command to execute on remote systems if not doing SCP",
                default_value="",
                parameter_group_info=[
                    ParameterGroupInfo(
                        group_name="run-command-plaintext-password"
                    ),
                    ParameterGroupInfo(
                        group_name="run-command-private-key"
                    )
                ]
            ),
        ]

    async def parse_arguments(self):
        self.load_args_from_json_string(self.command_line)

    async def parse_dictionary(self, dictionary):
        self.load_args_from_dictionary(dictionary)


class SSHAuthCommand(CommandBase):
    cmd = "sshauth"
    needs_admin = False
    help_cmd = "sshauth"
    description = """SSH to specified host(s) using the designated credentials. 
You can also use this to execute a specific command on the remote hosts via SSH or use it to SCP files.
"""
    version = 1
    author = "@xorrior"
    argument_class = SSHAuthArguments
    attackmapping = ["T1110"]
    browser_script = BrowserScript(script_name="sshauth", author="@djhohnstein")

    async def create_tasking(self, task: MythicTask) -> MythicTask:
        task.display_params = "Authenticate as " + task.args.get_arg("username")
        if "private-key" in task.args.get_parameter_group_name():
            task.display_params += " with a private key "
        else:
            task.display_params += " with password, " + task.args.get_arg("password") + ", "
        if "command" in task.args.get_parameter_group_name():
            task.display_params += "to run command, " + task.args.get_arg("command")
        elif "scp" in task.args.get_parameter_group_name():
            task.display_params += "to copy local file, " + task.args.get_arg("source") + ", to " + task.args.get_arg("destination")
        return task

    async def process_response(self, response: AgentResponse):
        pass

File: Payload_Type/poseidon/poseidon/agentfunctions/sshauth.go
package agentfunctions

import (
	"fmt"
	agentstructs "github.com/MythicMeta/MythicContainer/agent_structs"
	"path/filepath"
	"strings"
)

func init() {
	agentstructs.AllPayloadData.Get("poseidon").AddCommand(agentstructs.Command{
		Name: "sshauth",
		Description: `SSH to specified host(s) using the designated credentials. 
You can also use this to execute a specific command on the remote hosts via SSH or use it to SCP files.`,
		HelpString:          "sshauth",
		Version:             1,
		Author:              "@xorrior",
		MitreAttackMappings: []string{"T1110.003"},
		SupportedUIFeatures: []string{},
		CommandAttributes: agentstructs.CommandAttribute{
			SupportedOS: []string{},
		},
		AssociatedBrowserScript: &agentstructs.BrowserScript{
			ScriptPath: filepath.Join(".", "poseidon", "browserscripts", "sshauth_new.js"),
			Author:     "@djhohnstein",
		},
		CommandParameters: []agentstructs.CommandParameter{
			{
				Name:             "username",
				ModalDisplayName: "Username",
				Description:      "Authenticate to the designated hosts using this username",
				ParameterType:    agentstructs.COMMAND_PARAMETER_TYPE_STRING,
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: true,
						UIModalPosition:     1,
						GroupName:           "scp-private-key",
					},
					{
						ParameterIsRequired: true,
						UIModalPosition:     1,
						GroupName:           "scp-plaintext-password",
					},
					{
						ParameterIsRequired: true,
						UIModalPosition:     1,
						GroupName:           "scp-private-key-credstore",
					},
					{
						ParameterIsRequired: true,
						UIModalPosition:     1,
						GroupName:           "run-command-plaintext-password",
					},
					{
						ParameterIsRequired: true,
						UIModalPosition:     1,
						GroupName:           "run-command-private-key",
					},
					{
						ParameterIsRequired: true,
						UIModalPosition:     1,
						GroupName:           "run-command-private-key-credstore",
					},
				},
			},
			{
				Name:             "source",
				ModalDisplayName: "Source Filename",
				Description:      "If doing SCP, this is the source file",
				ParameterType:    agentstructs.COMMAND_PARAMETER_TYPE_STRING,
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: true,
						UIModalPosition:     6,
						GroupName:           "scp-private-key",
					},
					{
						ParameterIsRequired: true,
						UIModalPosition:     6,
						GroupName:           "scp-plaintext-password",
					},
					{
						ParameterIsRequired: true,
						UIModalPosition:     6,
						GroupName:           "scp-private-key-credstore",
					},
				},
			},
			{
				Name:             "destination",
				ModalDisplayName: "Destination Filename",
				Description:      "If doing SCP, this is the destination file",
				ParameterType:    agentstructs.COMMAND_PARAMETER_TYPE_STRING,
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: true,
						UIModalPosition:     7,
						GroupName:           "scp-private-key",
					},
					{
						ParameterIsRequired: true,
						UIModalPosition:     7,
						GroupName:           "scp-plaintext-password",
					},
					{
						ParameterIsRequired: true,
						UIModalPosition:     7,
						GroupName:           "scp-private-key-credstore",
					},
				},
			},
			{
				Name:             "private_key",
				ModalDisplayName: "Private Key Path On Target",
				ParameterType:    agentstructs.COMMAND_PARAMETER_TYPE_STRING,
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: true,
						UIModalPosition:     2,
						GroupName:           "scp-private-key",
					},
					{
						ParameterIsRequired: true,
						UIModalPosition:     2,
						GroupName:           "run-command-private-key",
					},
				},
			},
			{
				Name:                   "cred", // can't share `private_key` name
				ModalDisplayName:       "Credential",
				ParameterType:          agentstructs.COMMAND_PARAMETER_TYPE_CREDENTIAL,
				LimitCredentialsByType: []string{"plaintext", "key"},
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: true,
						UIModalPosition:     2,
						GroupName:           "scp-private-key-credstore",
					},
					{
						ParameterIsRequired: true,
						UIModalPosition:     2,
						GroupName:           "run-command-private-key-credstore",
					},
				},
			},
			{
				Name:             "port",
				ModalDisplayName: "SSH Port",
				Description:      "SSH Port if different than 22",
				DefaultValue:     22,
				ParameterType:    agentstructs.COMMAND_PARAMETER_TYPE_NUMBER,
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: false,
						UIModalPosition:     5,
						GroupName:           "scp-private-key",
					},
					{
						ParameterIsRequired: false,
						UIModalPosition:     5,
						GroupName:           "run-command-private-key",
					},
					{
						ParameterIsRequired: false,
						UIModalPosition:     5,
						GroupName:           "scp-private-key-credstore",
					},
					{
						ParameterIsRequired: false,
						UIModalPosition:     5,
						GroupName:           "scp-plaintext-password",
					},
					{
						ParameterIsRequired: false,
						UIModalPosition:     5,
						GroupName:           "run-command-plaintext-password",
					},
					{
						ParameterIsRequired: false,
						UIModalPosition:     5,
						GroupName:           "run-command-private-key-credstore",
					},
				},
			},
			{
				Name:             "password",
				ModalDisplayName: "Plaintext Password",
				Description:      "Authenticate to the designated hosts using this password",
				ParameterType:    agentstructs.COMMAND_PARAMETER_TYPE_STRING,
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: true,
						UIModalPosition:     3,
						GroupName:           "scp-plaintext-password",
					},
					{
						ParameterIsRequired: true,
						UIModalPosition:     3,
						GroupName:           "run-command-plaintext-password",
					},
				},
			},
			{
				Name:             "hosts",
				ModalDisplayName: "Array of CIDR notation for hosts",
				Description:      "Hosts that you will auth to",
				ParameterType:    agentstructs.COMMAND_PARAMETER_TYPE_ARRAY,
				DefaultValue:     []string{"127.0.0.1/32"},
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: true,
						UIModalPosition:     4,
						GroupName:           "scp-plaintext-password",
					},
					{
						ParameterIsRequired: true,
						UIModalPosition:     4,
						GroupName:           "scp-private-key-credstore",
					},
					{
						ParameterIsRequired: true,
						UIModalPosition:     4,
						GroupName:           "run-command-plaintext-password",
					},
					{
						ParameterIsRequired: true,
						UIModalPosition:     4,
						GroupName:           "run-command-private-key",
					},
					{
						ParameterIsRequired: true,
						UIModalPosition:     4,
						GroupName:           "scp-private-key",
					},
					{
						ParameterIsRequired: true,
						UIModalPosition:     4,
						GroupName:           "run-command-private-key-credstore",
					},
				},
			},
			{
				Name:             "command",
				ModalDisplayName: "Command to execute",
				Description:      "Command to execute on remote systems",
				ParameterType:    agentstructs.COMMAND_PARAMETER_TYPE_STRING,
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: true,
						UIModalPosition:     6,
						GroupName:           "run-command-plaintext-password",
					},
					{
						ParameterIsRequired: true,
						UIModalPosition:     6,
						GroupName:           "run-command-private-key",
					},
					{
						ParameterIsRequired: true,
						UIModalPosition:     6,
						GroupName:           "run-command-private-key-credstore",
					},
				},
			},
		},
		TaskFunctionCreateTasking: func(taskData *agentstructs.PTTaskMessageAllData) agentstructs.PTTaskCreateTaskingMessageResponse {
			response := agentstructs.PTTaskCreateTaskingMessageResponse{
				Success: true,
				TaskID:  taskData.Task.ID,
			}
			displayParams := ""
			username, err := taskData.Args.GetStringArg("username")
			if err != nil {
				response.Success = false
				response.Error = err.Error()
				return response
			}
			groupName, err := taskData.Args.GetParameterGroupName()
			if err != nil {
				response.Success = false
				response.Error = err.Error()
				return response
			}
			displayParams += fmt.Sprintf("as %s ", username)
			if strings.Contains(groupName, "private-key") {
				// authing with private key
				displayParams += fmt.Sprintf("with a private key")
			} else {
				// authing with plaintext password
				displayParams += fmt.Sprintf("with a plaintext password")
			}
			if strings.Contains(groupName, "command") {
				displayParams += fmt.Sprintf(" to run a command")
			} else {
				displayParams += fmt.Sprintf(" to copy a file")
			}
			response.DisplayParams = &displayParams

			return response
		},
		TaskFunctionParseArgDictionary: func(args *agentstructs.PTTaskMessageArgsData, input map[string]interface{}) error {
			return args.LoadArgsFromDictionary(input)
		},
		TaskFunctionParseArgString: func(args *agentstructs.PTTaskMessageArgsData, input string) error {
			return args.LoadArgsFromJSONString(input)
		},
	})
}
